pkg/image/layer.go

```diff
@@ -12,26 +14,65 @@ import (
 )
 
 type Layer struct {
-	area    sdl.Rect
-	buffer  *gfx.BufferArray
-	texture gfx.Texture
+	area      sdl.Rect
+	imgVAO    *gfx.VAO
+	selVAO    *gfx.VAO
+	texture   gfx.Texture
+	selTex    gfx.Texture
+	selDirty  bool
+	setupBuf  *gfx.BufferObject
+	offsetBuf *gfx.BufferObject
+	vertsBuf  *gfx.BufferObject
+	chunkSize int32
+	workers   int32
 }
 
-func NewLayer(offset sdl.Point, texture gfx.Texture) *Layer {
+func NewLayer(offset sdl.Point, texture gfx.Texture) (*Layer, error) {
+	selData := make([]byte, texture.GetWidth()*texture.GetHeight())
+	selTex, err := gfx.NewTexture(texture.GetWidth(), texture.GetHeight(), selData, gl.RED, 1, 1)
+	if err != nil {
+		return nil, err
+	}
+	var maxWorkers int32
+	gl.GetIntegeri_v(gl.MAX_COMPUTE_WORK_GROUP_COUNT, 0, &maxWorkers)
+	// TODO calculate intelligent chunk size on a per-image basis
+	chunkSize := int32(256)
+	workers := texture.GetWidth() * texture.GetHeight() / chunkSize
+	if workers > maxWorkers {
+		workers = maxWorkers
+	} else if workers == 0 {
+		workers = 1
+	}
+	log.Debugf("compute shader workers = %v", workers)
+
+	setupBuf := gfx.NewBufferObject()
+	setupBuf.BufferData(gl.SHADER_STORAGE_BUFFER, uint32(4*workers), gl.Ptr(nil), gl.DYNAMIC_READ)
+	offsetBuf := gfx.NewBufferObject()
+	offsetBuf.BufferData(gl.SHADER_STORAGE_BUFFER, uint32(4*(workers+1)), gl.Ptr(nil), gl.DYNAMIC_READ)
+	vertsBuf := gfx.NewBufferObject()
+	selVAO := gfx.NewVAO(gl.POINTS, []int32{2})
+	selVAO.SetVBO(vertsBuf)
+
 	return &Layer{
 		area: sdl.Rect{
 			X: offset.X,
 			Y: offset.Y,
 			W: texture.GetWidth(),
 			H: texture.GetHeight(),
 		},
-		buffer:  gfx.NewBufferArray(gl.TRIANGLES, []int32{2, 2}),
-		texture: texture,
-	}
+		imgVAO:    gfx.NewVAO(gl.TRIANGLES, []int32{2, 2}),
+		selVAO:    selVAO,
+		texture:   texture,
+		selTex:    selTex,
+		setupBuf:  setupBuf,
+		offsetBuf: offsetBuf,
+		vertsBuf:  vertsBuf,
+		chunkSize: chunkSize,
+		workers:   workers,
+	}, nil
 }
 
-// Render draws the ui.Component
-func (l Layer) Render(view sdl.FRect) {
+func (l Layer) Render(view sdl.FRect, program gfx.Program) {
```

Comment:
We should use pointer method receivers, so the method receiver type is the same for all the methods.
This method and `Destroy` still have value method receivers.